crates/sui-core/src/authority_aggregator.rs

```diff
@@ -1211,30 +1216,43 @@ where
                             Ok(VerifiedTransactionInfoResponse {
                                 signed_transaction: Some(inner_signed_transaction),
                                 ..
-                            }) if inner_signed_transaction.epoch() == self.committee.epoch => {
-                                let tx_digest = inner_signed_transaction.digest();
-                                debug!(tx_digest = ?tx_digest, name=?name.concise(), weight, "Received signed transaction from validator handle_transaction");
-                                state.signatures.push(inner_signed_transaction.into_inner().into_data_and_sig().1);
-                                state.good_stake += weight;
-                                if state.good_stake >= threshold {
-                                    self.metrics
-                                        .num_signatures
-                                        .observe(state.signatures.len() as f64);
-                                    self.metrics.num_good_stake.observe(state.good_stake as f64);
-                                    self.metrics.num_bad_stake.observe(state.bad_stake as f64);
-                                    state.certificate =
-                                        Some( CertifiedTransaction::new(
-                                            transaction_ref.data().clone(),
-                                            state.signatures.clone(),
-                                            &self.committee,
-                                        )?.verify(&self.committee)?);
+                            }) => {
+                                if inner_signed_transaction.epoch() != self.committee.epoch {
+                                    debug!(
+                                        tx_digest = ?tx_digest,
+                                        name=?name.concise(),
+                                        weight,
+                                        remote_epoch = inner_signed_transaction.epoch(),
+                                        local_epoch = self.committee.epoch,
+                                        "Received epoch-mismatched epoch signed transaction from validator handle_transaction"
+                                    );
+                                    state.errors.push(
+                                        SuiError::MismatchedEpochFromValidatorHandleTransaction {
```

Comment:
I am ok with introducing a new error. But it would be good if we don't end up printing this on the screen when running stress test at epoch boundary. (I tried to keep it as WrongEpoch error so that it could be filtered out during printing)

Reply:
the purpose of this PR is to separate `MismatchedEpochFromValidatorHandleTransaction` from `WrongEpoch`. The two `MismatchedEpochFromValidatorHandleTransaction` scenarios here are unlikely because 1. validator shouldn't return signed tx in a previous epoch and 2. validator should re-signs effects for previously committed tx. 

I'll be surprised if this is visible in stress, that'd mean some bugs in validator codes.

Also the `WrongEpoch` error shouldn't be visible after the refactor, as QD will retry them behind the scenes

Reply:
Here is what happens at epoch boundary:
1. Some validators have already finished reconfiguration and are now at epoch E + 1.
2. AuthorityAggregator's committee is still at epoch E. Transactions are sent to validators and coming back with signed tx at E + 1.
3. Stress already reconfigured and inserted the new committee to the committee store (hence the new committee is accessible).
4. The old (still running) AuthorityAggregator will pass SafeClient because the committee at new epoch is valid, but AuthorityAggregator will hit this case.

Reply:
This is fair. I'll change the code to distinguish whether remote epoch is bigger or local - If the remote is bigger (the case you describe above), then it yields WrongEpoch. Theoretically, if it has 2f+1 responses all in the same future epoch then it's tolerable, but it adds complexity to handle and it's much simpler to just fail with transient error and wait for next retry.